walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,83 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public static final int START_OF_DAY = TimeRange.getTimeInMinutes(0, 0);
+  public static final int END_OF_DAY = TimeRange.getTimeInMinutes(23, 59);
+
+  public static final Comparator<Event> ORDER_BY_START =
+      new Comparator<Event>() {
+        @Override
+        public int compare(Event a, Event b) {
+          return Long.compare(a.getWhen().start(), b.getWhen().start());
+        }
+      };
+
+  public Collection<TimeRange> query(Collection<Event> eventsCollection, MeetingRequest request) {
+    HashSet<String> requestAttendees = new HashSet<String>(request.getAttendees());
+    ArrayList<Event> events =
+        removeIrrelevantEvents(requestAttendees, new ArrayList<Event>(eventsCollection));
+    List<TimeRange> res = new ArrayList<TimeRange>();
+    if (events.isEmpty()) {
+      addIfValid(TimeRange.fromStartEnd(0, END_OF_DAY, true), res, request.getDuration());
+      return res;
+    }
+
+    Collections.sort(events, ORDER_BY_START);
+
+    // Add first gap.
+    addIfValid(
+        TimeRange.fromStartEnd(0, events.get(0).getWhen().start(), false),
+        res,
+        request.getDuration());
+    int end = events.get(0).getWhen().end();
+    for (Event event : events) {
+      // event can be merged with current time range
+      if (event.getWhen().start() <= end) {
+        end = Math.max(end, event.getWhen().end());
```

Comment:
Optional:

I'm keen on `continue` in these contexts to eliminate extra indentation. It's similar to guard clauses in functions that return early.

```java
    for (Event event : events) {
      // event can be merged with current time range
      if (event.getWhen().start() <= end) {
        end = Math.max(end, event.getWhen().end());
        continue
      }
      // Add the time range we were tracking, start a new one from event.
      addIfValid(
          TimeRange.fromStartEnd(end, event.getWhen().start(), false),
          res,
          request.getDuration());
      end = event.getWhen().end();
    }
```